Add NSFS monitoring & refactor NS monitoring

liranmauda opened the pull request:
### Explain the changes
1. Adding BG worker to the endpoint and running the `namespace_monitor` on it
2. Refractor `namespace_monitor`
3. Add the ability to decide if we should monitor a namespace or not (passing a `should_monitor` function to the `namespace_monitor` constructor.
4. Writing and deleting a small file into the `fs_root_path`
5. Reporting on a monitor fail and on an I/O fail
6. Adding namespace_resource_id to the namespaceFS constructor, and will not report if it is undefined.

guymguym commented on src/sdk/namespace_fs.js:
```diff
@@ -532,6 +547,17 @@ class NamespaceFS {
                 await this._upload_stream(params.source_stream, upload_path, fs_account_config);
             }
         } catch (err) {
+            try {
+                console.log('LMLM before upload_object report');
+                object_sdk.rpc_client.pool.update_issues_report({
+                    namespace_resource_id: this.namespace_resource_id,
+                    error_code: err.code,
+                    time: Date.now(),
+                });
+                console.log('LMLM after upload_object report');
+            } catch (e) {
+                console.log('update_issues_report on error:', e, 'ignoring.');
+            }
```
1. Refactor to a function - we don't want to repeat this boilerplate everytime.
2. Lets filter to report only on errors with `err.code`

guymguym commented on src/server/bg_services/namespace_fs_monitor.js:
```diff
@@ -0,0 +1,86 @@
+const auth_server = require('../common_services/auth_server');
+const system_store = require('../system_services/system_store').get_instance();
+
+class NamespaceFSMonitor {
```
Why is this separated from NamespaceMonitor?

liranmauda replied:
It felt more right, As this Monitor is only on the endpoint and not on the core (at list for now).
Another thing is that the `requires` are different and the test also.
The flow is different as we don't even need a connection.

guymguym replied:
To me it makes more sense to have it as another case under the same namespace monitor class, rather than duplicating the entire class and add if(nsfs) and if(!nsfs) in each one respectfully. Sounds like you could extend the existing class with another function and that's it.

liranmauda replied:
It is not just namespace vs nsfs, It is where it runs.
We are running it on the endpoint, and if we will run the Namespace_monitor on the endpoint it will run per namespace, even if it is not nsfs.
of course, we can also protect that, but I find it cleaner to divide into 2 files, one that runs in core and checks namespace and the other that's run in the endpoint and checks nsfs. 
WDYT?

guymguym commented on src/sdk/object_sdk.js:
```diff
@@ -180,8 +180,13 @@ class ObjectSDK {
         try {
             // NAMESPACE_FS HACK
             if (process.env.NAMESPACE_FS) {
+                const namespace_resource_id = _.extend({}, bucket.namespace.write_resource).resource.id;
```
I prefer more explicit and not to make an object just for checking existence.
We can start using the new `?.` operator for these cases:
```suggestion
                const namespace_resource_id = bucket.namespace.write_resource?.resource.id;
```

guymguym commented on src/core/nsfs.js:
```diff
@@ -104,7 +104,11 @@ async function main(argv = minimist(process.argv.slice(2))) {
+            const ns_fs = new NamespaceFS({
+                bucket_path: fs_root + '/' + bucket_name,
+                bucket_id: '000000000000000000000000',
+                namespace_resource_id: '9999'
```
can we pass undefined `namespace_resource_id` and then namespace_fs will avoid sending reports?
because in standalone more there is nowhere to send it.

liranmauda committed:
Monitoring namspace FS

Add Monitoring for namspace FS:
- Add a bg worker for the endpoint
- Refractor namespace monitor
- Add Monitoring on NSFS I/O fails

Signed-off-by: liranmauda <liran.mauda@gmail.com>

jeniawhite commented on src/server/bg_services/namespace_monitor.js:
```diff
@@ -46,96 +49,140 @@ class NamespaceMonitor {
+    }
+
+    async test_single_namespace_fs_resource_validity(nsr) {
+        const file = path.join(nsr.nsfs_config.fs_root_path, 'create_object_upload');
```
Do we actually want to create an object with the name `create_object_upload` under the root path of the FS?
Doesn't seem like something reasonable to do.
Even if we want to create an object I would expect it to be with the `.` prefix so it won't be visible by the FS and definitely not with this name.
@guymguym

guymguym replied:
Right, but I don't know why are we writing in the first place - why not just stat the fs_root_path?
What else is missing? detecting no-space? maybe we should just use statfs?
https://man7.org/linux/man-pages/man2/statfs.2.html

liranmauda replied:
We thought of having a hidden directory for this, but as we create and delete the file we agreed that we will not create a directory but write the file into the `fs_root_path`.
With that said we can defiantly create a hidden file (and change the name).
@guymguym @nimrod-becker @jeniawhite WDYT?

liranmauda replied:
`stat` is not working properly.
for example, when you don't have the directory in your file system (it was deleted directly), you will still have this directory on the endpoint pod. in this case, `stat` will return an answer instead of an error that we expect.

guymguym replied:
But why do you think that writing/deleting/whatever will fail in case of an unmounted dir?
I don't think we are really trying to verify that the dir is mounted.
That's for kubernetes to do.

guymguym replied:
I would still avoid writing and use stat/statfs instead.

liranmauda replied:
> But why do you think that writing/deleting/whatever will fail in case of an unmounted dir?

Per my tests, I am deleting the fs_root_path in the minikube, and write fails with `ENOENT`
`stat` did not fail. 

So at list, in the case I checked `stat` is not monitoring and `write` does.

liranmauda replied:
> I would still avoid writing and use stat/statfs instead.

What about `unkink` without `write`?  WDYT?

guymguym replied:
OK can you give details on your test - what type of PVC did you use - hostpath? what did you delete in minikube exactly? are you saying that stat() succeeded on a missing dir??

Reminding of the problems we want to identify -

Operator:
- PVC not ready / mount failure - Update nsstore status to rejected. If possible, remove  from endpoints to keep serving other resources.

Filesystem:
- No/low space - ENOSPC / statfs
- No permissions - EPERM / EACCESS
- Not exists - ENOENT / EBADF? / ENOTFOUND?

So how are we doing in identifying all those?

jeniawhite commented on src/server/bg_services/namespace_monitor.js:
```diff
@@ -46,96 +49,140 @@ class NamespaceMonitor {
-            }
-            default:
-                dbg.error('namespace_monitor: invalid endpoint type', nsr.endpoint_type);
+            return true;
```
Maybe it is better to print the resource as well as we do below for other types?

guymguym commented on src/sdk/namespace_fs.js:
```diff
@@ -162,6 +164,20 @@ class NamespaceFS {
 
+    run_update_issues_report(object_sdk, err) {
+        if (!err.code) return;
+        if (!this.namespace_resource_id) return;
```
@jeniawhite - This is for standalone nsfs where we don't have a resource id, and do not even have the pool server to send the update_issues_report api call. So when namespace_resource_id is undefined, we skip reporting.

@liranmauda - You should definitely add comments to explain this in the constructor, and also here.

liranmauda replied:
This is per @guymguym request https://github.com/noobaa/noobaa-core/pull/6549#discussion_r646101060
In `nsfs.js` which is standalone, we want to avoid the report.

guymguym replied:
Add a comment to explain that in the code!

guymguym commented on src/server/bg_services/namespace_monitor.js:
```diff
@@ -46,96 +49,140 @@ class NamespaceMonitor {
+        });
+    }
+
+    init_nsr_s3_connection_to_target(nsr) {
```
I would just embed this function content directly inside `test_s3_resource()`.

guymguym commented on src/server/bg_services/namespace_monitor.js:
```diff
@@ -46,96 +49,140 @@ class NamespaceMonitor {
+        this.nsr_connections_obj[nsr._id] = conn;
+    }
+
+    init_nsr_blob_connection_to_target(nsr) {
```
I would just embed this function content directly inside `test_blob_resource()`.

guymguym commented on src/server/bg_services/namespace_monitor.js:
```diff
@@ -46,96 +49,140 @@ class NamespaceMonitor {
-            return;
-        }
-        const { endpoint_type, target_bucket} = nsr_info.connection;
+    async test_single_namespace_s3_resource_validity(nsr) {
```
Rename - I would drop extra words that don't add meaning:
```suggestion
    async test_s3_resource(nsr) {
```

guymguym commented on src/server/bg_services/namespace_monitor.js:
```diff
@@ -46,96 +49,140 @@ class NamespaceMonitor {
+
+    }
+
+    async test_single_namespace_blob_resource_validity(nsr) {
```
Rename - I would drop extra words that don't add meaning:
```suggestion
    async test_blob_resource(nsr) {
```

guymguym commented on src/server/bg_services/namespace_monitor.js:
```diff
@@ -46,96 +49,140 @@ class NamespaceMonitor {
+        }
+    }
+
+    async test_single_namespace_fs_resource_validity(nsr) {
```
Rename - I would drop extra words that don't add meaning:
```suggestion
    async test_nsfs_resource(nsr) {
```

guymguym commented on src/server/bg_workers.js:
```diff
@@ -92,6 +92,13 @@ function remove_master_workers() {
 
 }
 
+function should_monitor(nsr) {
```
unneeded as top level function, this is a small arrow func

guymguym commented on src/endpoint/endpoint.js:
```diff
@@ -138,6 +147,13 @@ async function start_endpoint(options = {}) {
     }
 }
 
+function should_monitor(nsr) {
```
unneeded as top level function, this is a small arrow func

guymguym commented on src/server/bg_services/namespace_monitor.js:
```diff
@@ -7,17 +7,20 @@ const auth_server = require('../common_services/auth_server');
 class NamespaceMonitor {
 
-    constructor({ name, client}) {
+    constructor({ name, client, should_monitor }) {
```
Use jsdoc!
```suggestion
    /**
     * @param {{
     *   name: string;
     *   client: nb.APIClient;
     *   should_monitor: (nsr: nb.NamespaceResource) => boolean;
     * }} params
     */
    constructor({ name, client, should_monitor }) {
```

guymguym commented on src/sdk/namespace_fs.js:
```diff
@@ -121,13 +121,15 @@ class NamespaceFS {
      *  bucket_path: string;
      *  fs_backend?: string;
      *  bucket_id: string;
+     *  namespace_resource_id: string;
```
Since namespace_resource_id is optional you want the jsdoc to represent that -
```suggestion
     *  namespace_resource_id?: string;
```

guymguym commented on src/sdk/namespace_fs.js:
```diff
@@ -516,8 +533,8 @@ class NamespaceFS {
         // console.log('NamespaceFS.upload_object:', upload_path, '->', file_path);
-        await Promise.all([this._make_path_dirs(file_path, fs_account_config), this._make_path_dirs(upload_path, fs_account_config)]);
         try {
+            await Promise.all([this._make_path_dirs(file_path, fs_account_config), this._make_path_dirs(upload_path, fs_account_config)]);
```
Break this line
```suggestion
            await Promise.all([
                this._make_path_dirs(file_path, fs_account_config), 
                this._make_path_dirs(upload_path, fs_account_config)
            ]);
```

liranmauda commented on src/endpoint/endpoint.js:
```diff
@@ -354,6 +370,19 @@ function setup_http_server(server) {
     // });
 }
 
+function register_bg_endpoint_worker(worker, run_batch_function) {
```
True, but as loading it requires not to `register_common_services` twice, not `register_bg_services`, and maybe remove the `md_server.js` completely, We should not do it now.

guymguym replied:
I suggest we move the register_bg_worker function from -
https://github.com/noobaa/noobaa-core/blob/3f34c1a7389e68f1f20967e046128545a28995ca/src/server/bg_workers.js#L76-L86
into the utility module in -
https://github.com/noobaa/noobaa-core/blob/master/src/util/background_scheduler.js
And then share it from both.
Can work?

guymguym commented on src/server/bg_services/namespace_monitor.js:
```diff
@@ -46,96 +55,120 @@ class NamespaceMonitor {
-                dbg.error('namespace_monitor: invalid endpoint type', nsr.endpoint_type);
-        }
-        if (conn) {
+        if (this.nsr_connections_obj[nsr._id]) {
+            conn = this.nsr_connections_obj[nsr._id];
+        } else {
```
```suggestion
        let conn = this.nsr_connections_obj[nsr._id];
        if (!conn) {
```

guymguym reviewed (changes requested):
Some more comments. The biggest thing left from my point of view is that writing to the FS is not really needed and we better avoid it in the monitoring flow IMHO.

guymguym commented on src/test/unit_tests/test_namespace_fs.js:
```diff
@@ -35,8 +35,8 @@ mocha.describe('namespace_fs', function() {
 
-    const ns_src = new NamespaceFS({ bucket_path: ns_src_bucket_path, bucket_id: '1'});
-    const ns_tmp = new NamespaceFS({ bucket_path: ns_tmp_bucket_path, bucket_id: '2'});
+    const ns_src = new NamespaceFS({ bucket_path: ns_src_bucket_path, bucket_id: '1', namespace_resource_id: '1' });
```
why are we setting namespace_resource_id here? 
does it work to report issues on this fake id in the test?

liranmauda replied:
true, I will change it explicitly to `namespace_resource_id: undefined`

guymguym commented on src/server/bg_workers.js:
```diff
@@ -72,18 +71,7 @@ function register_rpc() {
-    }
-    background_scheduler.run_background_worker(worker);
-}
+const register_bg_worker = background_scheduler.register_bg_worker;
```
this will not work. this function is not bound to its instance so calling it will fail to resolve `this`.
You could replace with -
```suggestion
const register_bg_worker = 
    (worker, run_batch_function) => 
    background_scheduler.register_bg_worker(worker, run_batch_function)
```

jeniawhite commented on src/server/bg_workers.js:
```diff
@@ -72,18 +71,9 @@ function register_rpc() {
-    }
-    background_scheduler.run_background_worker(worker);
-}
+const register_bg_worker =
```
I don't understand the point of this transformation.
You basically create a function to point to a function without doing any manipulation.
Why not just call the function directly?

liranmauda replied:
https://github.com/noobaa/noobaa-core/pull/6549#discussion_r649153223

liranmauda replied:
+ We want to avoid refractoring all the 21 places from `register_bg_worker` to `background_scheduler.register_bg_worker`

jeniawhite commented on src/server/bg_services/namespace_monitor.js:
```diff
@@ -46,96 +54,116 @@ class NamespaceMonitor {
-                    this.init_nsr_connection_to_target(nsr);
+                if (endpoint_type === 'NSFS') {
+                    await this.test_nsfs_resource(nsr);
+                } else if (['AWS', 'S3_COMPATIBLE', 'IBM_COS'].includes(endpoint_type)) {
```
In the case of non-NSFS, the connection holds the type inside the endpoint_type variable.
It is an object, not a type.
Just take the endpoint_type from the connection above when you assign.
